docs/code/MlNetCookBook.md

```diff
@@ -0,0 +1,212 @@
+# ML.NET Cookbook
+
+This document is intended to provide essential samples for common usage patterns of ML.NET. 
+It is advisable to be at least minimally familiar with [high-level concepts of ML.NET](MlNetHighLevelConcepts.md), otherwise the terminology in this document may be foreign to you.
+
+## How to use this cookbook
+
+Developers often work by copying and pasting source code from somewhere and then adapting it to their needs. We do it all the time.
+
+So, we decided to embrace the pattern and provide an authoritative set of example usages of ML.NET, for many common scenarios that you may encounter.
+These examples are multi-purpose:
+
+- They can kickstart your development, so that you don't start from nothing,
+- They are annotated and verbose, so you have easier time adapting them to your needs.
+
+Each sample also contains a snippet of the data file used in the sample. We mostly use snippets from our test datasets for that.
+
+Please feel free to search this page and use any code that suits your needs.
+
+### List of recipes
+
+- [How do I load data from a text file?](#how-do-i-load-data-from-a-text-file)
+- [How do I load data with many columns from a CSV?](#how-do-i-load-data-with-many-columns-from-a-csv)
+- [How do I look at the intermediate data?](#how-do-i-look-at-the-intermediate-data)
+
+## How do I load data from a text file?
+
+`TextLoader` is used to load data from text files. You will need to specify what are the data columns, what are their types, and where to find them in the text file. 
+
+Note that it's perfectly acceptable to read only some columns of a file, or read the same column multiple times.
+
+Example file (https://github.com/dotnet/machinelearning/blob/master/test/data/adult.tiny.with-schema.txt):
+```
+Label	Workclass	education	marital-status
+0	Private	11th	Never-married
+0	Private	HS-grad	Married-civ-spouse
+1	Local-gov	Assoc-acdm	Married-civ-spouse
+1	Private	Some-college	Married-civ-spouse
+
+```
+
+This is how you can read this data:
+```c#
+// Create a new environment for ML.NET operations. It can be used for exception tracking and logging, 
+// as well as the source of randomness.
+var env = new LocalEnvironment();
+
+// Create the reader: define the data columns and where to find them in the text file.
+var reader = TextLoader.CreateReader(env, ctx => (
+        // A boolean column depicting the 'target label'.
+        IsOver50K: ctx.LoadBool(0),
+        // Three text columns.
+        Workclass: ctx.LoadText(1),
+        Education: ctx.LoadText(2),
+        MaritalStatus: ctx.LoadText(3)),
+    hasHeader: true);
+
+// Now read the file.
+var data = reader.Read(new MultiFileSource(dataPath));
+```
+
+If the schema of the data is not known at compile time, or too cumbersome, you can revert to the dynamically-typed API: 
```

Comment:
It's not really clear what has changed here, beyond the syntax. This example still shows that we know the column names, types, and column indices. 
I realize this is because we want the example to be simple, but how do we convey why this approach might be needed?

Reply:
(after reading further): Ah this information might be provided programmatically. Ignore this.

One thing that might be useful is showing how to read all remaining columns.

Reply:
not sure I get it.

---
In reply to: [219395834](https://github.com/dotnet/machinelearning/pull/960#discussion_r219395834) [](ancestors = 219395834)

Reply:
"how to read all remaining columns": if my file has a label in the first column and numeric features in all remaining columns, can I do something to the effect of `c.LoadFloat(1,*)`? As opposed to finding how many numeric columns I have.

Reply:
I see now

---
In reply to: [221339003](https://github.com/dotnet/machinelearning/pull/960#discussion_r221339003) [](ancestors = 221339003)

Reply:
Maybe later.

---
In reply to: [221377364](https://github.com/dotnet/machinelearning/pull/960#discussion_r221377364) [](ancestors = 221377364,221339003)